django/thunderstore/repository/views/team_settings.py

```diff
@@ -111,17 +112,32 @@ def get_context_data(self, **kwargs):
         return context
 
     def form_invalid(self, form):
-        messages.error(
-            self.request, "There was a problem performing the requested action"
-        )
+        error_msg = "There was a problem performing the requested action"
+        messages.error(self.request, error_msg)
         capture_exception(ValidationError(form.errors))
         return super().form_invalid(form)
 
     def form_valid(self, form):
+        if "demote" in self.request.POST or "promote" in self.request.POST:
+            return self._handle_demote_and_promote(form)
+
         form.save()
         messages.success(self.request, "Action performed successfully")
         return redirect(self.object.settings_url)
 
+    def _handle_demote_and_promote(self, form):
+        try:
+            update_team_member(
+                agent=self.request.user,
+                team_member=form.instance,
+                role=form.cleaned_data["role"],
+            )
+        except ValidationError as e:
+            return self.form_invalid(form)
+
+        messages.success(self.request, "Action performed successfully")
+        return redirect(self.object.settings_url)
```

Comment:
I feel like instead of editing the view here you should edit the form's success action to call the service layer, complicates the view less and encapsulates the logic better. This likely also applies to https://github.com/thunderstore-io/Thunderstore/pull/1116, I just didn't realize to consider it since the view was more straight forward.